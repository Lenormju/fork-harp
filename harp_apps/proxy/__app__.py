"""
Proxy Application

"""

from .lifecycle import ProxyLifecycle
from .settings import ProxySettings


class ProxyApplication:
    Settings = ProxySettings
    Lifecycle = ProxyLifecycle
