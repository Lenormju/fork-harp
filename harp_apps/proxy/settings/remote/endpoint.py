from typing import Annotated, FrozenSet, Set

from pydantic import Field, HttpUrl, field_serializer, field_validator

from harp.config import Configurable
from harp_apps.proxy.constants import AVAILABLE_POOLS


class RemoteEndpointSettings(Configurable):
    """
    A ``HttpEndpoint`` is an instrumented target URL that a proxy will use to route requests. It is used as the
    configuration parser for ``proxy.endpoints[].remote.endpoints[]`` settings.

    .. code-block:: yaml

        url: "http://my-endpoint:8080"
        pools: [fallback]  # omit for default pool
        failure_threshold: 3
        success_threshold: 1
    """

    url: HttpUrl
    pools: Set[str] = {"default"}
    failure_threshold: Annotated[int, Field(gt=0)] = 1
    success_threshold: Annotated[int, Field(gt=0)] = 1

    @field_validator("pools")
    @classmethod
    def validate_pools(cls, pools: FrozenSet[str]) -> FrozenSet[str]:
        if not pools.issubset(AVAILABLE_POOLS):
            raise ValueError(f"Invalid pool names: {', '.join(pools.difference(AVAILABLE_POOLS))}.")
        return pools

    @field_serializer("pools", when_used="json")
    @classmethod
    def serialize_in_order(cls, value: Set[str]):
        return sorted(value)
