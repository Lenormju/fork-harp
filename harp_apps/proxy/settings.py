from dataclasses import field
from typing import Optional

from harp.config import Settings, asdict, settings_dataclass
from harp_apps.proxy.models.remotes import HttpRemote


@settings_dataclass
class ProxySettings(Settings):
    """
    Configuration parser for ``proxy`` settings.

    .. code-block:: yaml

        endpoints:
          # see ProxyEndpoint
          - ...

    """

    endpoints: list["ProxyEndpoint"] | None = field(default_factory=list)

    def __post_init__(self):
        if self.endpoints is None:
            self.endpoints = []

        self.endpoints = [
            endpoint if isinstance(endpoint, ProxyEndpoint) else ProxyEndpoint(**endpoint)
            for endpoint in self.endpoints
        ]


@settings_dataclass
class ProxyEndpoint(Settings):
    """
    Configuration parser for ``proxy.endpoints[]`` settings.

    .. code-block:: yaml

        name: my-endpoint
        port: 8080
        description: My endpoint
        remote:
          # see HttpRemote
          ...

    A shorthand syntax is also available for cases where you only need to proxy to a single URL and do not require
    fine-tuning the endpoint settings:

    .. code-block:: yaml

        name: my-endpoint
        port: 8080
        description: My endpoint
        url: http://my-endpoint:8080

    """

    name: str
    port: int
    description: Optional[str] = None

    # for backward compatibility, "short syntax"
    url: str | None = None

    # resilience-compatible remote definition, with url pools, probes, etc.
    remote: HttpRemote | None = None

    def __post_init__(self):
        if self.remote is not None and self.url is not None:
            raise ValueError(
                "You can't define both proxy.endpoints[].remote and proxy.endpoints[].url, the second one is just a "
                "shorthand syntax for the first one."
            )

        if self.url is not None:
            self.remote = HttpRemote([self.url])

        if isinstance(self.remote, dict):
            self.remote = HttpRemote(**self.remote)

        if not isinstance(self.remote, HttpRemote):
            raise ValueError(f"Invalid remote configuration: {self.remote}")

    def _asdict(self, /, *, secure=True):
        return {
            "name": self.name,
            "port": self.port,
            "description": self.description,
            "remote": asdict(self.remote, secure=secure),
        }
