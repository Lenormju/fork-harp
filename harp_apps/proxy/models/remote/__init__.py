import asyncio
import warnings
from collections import deque
from functools import cached_property
from typing import Mapping

import httpx

from harp import get_logger
from harp.config import StatefulConfigurableWrapper
from harp.utils.urls import normalize_url
from harp_apps.proxy.constants import CHECKING, DEFAULT_POOL, DOWN, FALLBACK_POOL, UP
from harp_apps.proxy.settings.remote import RemoteSettings

from .endpoint import RemoteEndpoint
from .probe import RemoteProbe

logger = get_logger(__name__)


class Remote(StatefulConfigurableWrapper[RemoteSettings]):
    #: Current pool deque contains the list of available URLs, from least recently used to most recently used. It will
    #: be rotated on each request to implement a naive round-robin strategy.
    current_pool: deque[RemoteEndpoint]

    #: Remote endpoints with current status.
    endpoints: Mapping[str, RemoteEndpoint]

    def __init__(self, settings: RemoteSettings):
        super().__init__(settings)

        self.endpoints = {
            normalize_url(str(endpoint.url)): RemoteEndpoint(endpoint) for endpoint in (settings.endpoints or ())
        }
        self.current_pool = deque()

        self.refresh()

    def __getitem__(self, url: str) -> RemoteEndpoint:
        return self.endpoints[normalize_url(url)]

    @cached_property
    def probe(self):
        return RemoteProbe(self.settings.probe)

    def refresh(self):
        """Recompute the current pool of endpoints."""
        refreshed: deque[RemoteEndpoint] = deque()
        for endpoint in self.endpoints.values():
            if DEFAULT_POOL in endpoint.pools and endpoint.status >= CHECKING:
                refreshed.append(endpoint)

        if len(refreshed) < self.min_pool_size:
            for endpoint in self.endpoints.values():
                if FALLBACK_POOL in endpoint.pools and endpoint.status >= CHECKING:
                    self.current_pool_name = FALLBACK_POOL
                    refreshed.append(endpoint)
        else:
            self.current_pool_name = DEFAULT_POOL

        self.current_pool = refreshed

    def get_url(self) -> str:
        """Get next candidate url from the current pool, then rotate."""
        try:
            return str(self.current_pool[0].url)
        except IndexError as exc:
            raise IndexError("No available URLs for remote.") from exc

        finally:
            self.current_pool.rotate(-1)

    def notify_url_status(self, url, status):
        """
        Take into account a http status code to update the status of the url. Behaviour differs depending on the
        instance configuration (some considers 5xx as down, some ignores non-network errors).
        """
        if ("http_5xx" in self.break_on and 500 <= status < 600) or (
            "http_4xx" in self.break_on and 400 <= status < 500
        ):
            if self[url].failure(f"HTTP_{status}"):
                self.refresh()

    def set_down(self, url):
        old_status = self[url].status

        self[url].status = DOWN

        if old_status >= DOWN:

            async def delayed_set_checking():
                await asyncio.sleep(self.delay_for_checking_down_endpoints)
                if self[url].status == DOWN:
                    self.set_checking(url)
                del self[url].delayed_set_checking

            try:
                self[url].delayed_set_checking = asyncio.create_task(delayed_set_checking())
            except RuntimeError as exc:
                warnings.warn(f"Failed to schedule delayed checking state: {exc}")

        self.refresh()

    def set_checking(self, url):
        self[url].status = CHECKING
        self.refresh()

    def set_up(self, url):
        self[url].status = UP
        self.refresh()

    async def check(self):
        """Uses the probe (luke), to check the health of each urls. It is also done on fallback and inactive urls, to
        ensure that they are ready in case we need them."""
        if self.probe is None:
            return

        state_changed = False

        async with httpx.AsyncClient(verify=self.probe.verify) as client:
            for url in self.endpoints.values():
                url_state_changed = await self.probe.check(client, url)
                state_changed |= url_state_changed

        if state_changed:
            self.refresh()

    async def check_forever(self):
        while True:
            try:
                await self.check()
            except Exception as exc:
                logger.error(f"Failed to check remote health: {exc}")
            await asyncio.sleep(self.probe.interval)
