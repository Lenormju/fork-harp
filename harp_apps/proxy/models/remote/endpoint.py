from typing import Optional

from harp.config import StatefulConfigurableWrapper
from harp_apps.proxy.constants import CHECKING, DOWN, UP
from harp_apps.proxy.settings.remote import RemoteEndpointSettings


class RemoteEndpoint(StatefulConfigurableWrapper[RemoteEndpointSettings]):
    status: int
    failure_score: int
    success_score: int
    failure_reasons: Optional[set] = None

    def __init__(self, settings: RemoteEndpointSettings):
        super().__init__(settings)

        self.status = CHECKING
        self.failure_score = 0
        self.success_score = 0

    def success(self):
        """Returns a boolean indicating if a state change happened."""
        self.failure_score = 0
        self.success_score += 1

        if self.success_score >= self.success_threshold:
            if self.status != UP:
                self.failure_reasons = None
                self.status = UP
                return True
        else:
            if self.status != CHECKING:
                self.status = CHECKING
                return True

        return False

    def failure(self, reason: str = None):
        """Returns a boolean indicating if a state change happened."""
        self.success_score = 0
        self.failure_score += 1

        if self.failure_reasons is None:
            self.failure_reasons = set()

        if reason:
            self.failure_reasons.add(reason)

        if self.failure_score >= self.failure_threshold:
            if self.status != DOWN:
                self.status = DOWN
                return True
        else:
            if self.status != CHECKING:
                self.status = CHECKING
                return True

        return False


"""
    status: int = CHECKING

    def __post_init__(self):
        self.url = normalize_url(self.url)
        self.pools = set(self.pools)
        unknown_pools = self.pools.difference({DEFAULT_POOL, FALLBACK_POOL})
        if len(unknown_pools):
            raise ValueError(f"Invalid pool names: {unknown_pools}")


    def _asdict(self, /, *, secure=True, with_status=False):
        return {
            "url": self.url,
            **(
                {"failure_threshold": self.failure_threshold}
                if self.failure_threshold != type(self).failure_threshold
                else {}
            ),
            **(
                {"success_threshold": self.success_threshold}
                if self.success_threshold != type(self).success_threshold
                else {}
            ),
            **({"pools": list(self.pools)} if len(self.pools) > 0 else {}),
            **({"status": humanize_status(self.status)} if with_status else {}),
            **({"failure_reasons": list(sorted(self.failure_reasons))} if with_status and self.failure_reasons else {}),
        }
"""
