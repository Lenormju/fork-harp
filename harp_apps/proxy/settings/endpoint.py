from typing import Optional

from pydantic import model_validator

from harp.config import Configurable
from harp_apps.proxy.settings.remote import RemoteEndpointSettings, RemoteSettings


class EndpointSettings(Configurable):
    """
    Configuration parser for ``proxy.endpoints[]`` settings.

    .. code-block:: yaml

        name: my-endpoint
        port: 8080
        description: My endpoint
        remote:
          # see HttpRemote
          ...

    A shorthand syntax is also available for cases where you only need to proxy to a single URL and do not require
    fine-tuning the endpoint settings:

    .. code-block:: yaml

        name: my-endpoint
        port: 8080
        description: My endpoint
        url: http://my-endpoint:8080

    """

    name: str
    port: int
    description: Optional[str] = None

    # for backward compatibility, "short syntax"
    url: Optional[str] = None

    # resilience-compatible remote definition, with url pools, probes, etc.
    remote: Optional[RemoteSettings] = None

    @model_validator(mode="before")
    @classmethod
    def convert_old_school_urls_to_remote(cls, values):
        if "url" in values and values["url"] is not None:
            if "remote" in values and values["remote"] is not None:
                raise ValueError(
                    "You can't define both proxy.endpoints[].remote and proxy.endpoints[].url, the second one is just "
                    "a historical shorthand syntax for the first one."
                )
            values["remote"] = RemoteSettings(endpoints=[RemoteEndpointSettings(url=values.pop("url"))])
        return values


"""
    def __post_init__(self):
        if self.remote is not None and self.url is not None:
            raise ValueError(
                "You can't define both proxy.endpoints[].remote and proxy.endpoints[].url, the second one is just a "
                "shorthand syntax for the first one."
            )

        if self.url is not None:
            self.remote = HttpRemote([self.url])

        if isinstance(self.remote, dict):
            self.remote = HttpRemote(**self.remote)

        if not isinstance(self.remote, HttpRemote):
            raise ValueError(f"Invalid remote configuration: {self.remote}")

    def _asdict(self, /, *, secure=True, with_status=False):
        return {
            "name": self.name,
            "port": self.port,
            "description": self.description,
            "remote": self.remote._asdict(secure=secure, with_status=with_status),
        }
"""
