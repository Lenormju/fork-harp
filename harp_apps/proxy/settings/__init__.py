from harp.config import Configurable, Stateful

from .endpoint import EndpointSettings
from .remote import Remote, RemoteEndpoint, RemoteEndpointSettings, RemoteProbe, RemoteProbeSettings, RemoteSettings

__all__ = [
    "Proxy",
    "ProxySettings",
    "Remote",
    "RemoteEndpoint",
    "RemoteEndpointSettings",
    "RemoteProbe",
    "RemoteProbeSettings",
    "RemoteSettings",
]


class BaseProxySettings(Configurable):
    pass


class ProxySettings(BaseProxySettings):
    """
    Configuration parser for ``proxy`` settings.

    .. code-block:: yaml

        endpoints:
          # see ProxyEndpoint
          - ...

    """

    endpoints: list[EndpointSettings] = []


class Proxy(Stateful[ProxySettings]):
    pass
