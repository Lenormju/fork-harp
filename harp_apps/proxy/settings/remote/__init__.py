from decimal import Decimal
from typing import Iterable, Optional

from pydantic import field_serializer, field_validator

from harp.config import Configurable
from harp.utils.urls import normalize_url
from harp_apps.proxy.constants import ALL_BREAK_ON_VALUES, BREAK_ON_NETWORK_ERROR, BREAK_ON_UNHANDLED_EXCEPTION
from harp_apps.proxy.settings.remote.probe import RemoteProbeSettings

from .endpoint import RemoteEndpointSettings


class RemoteSettings(Configurable):
    """
    A ``HttpRemote`` is a collection of endpoints that a proxy will use to route requests. It is used as the
    configuration parser for ``proxy.endpoints[].remote`` settings.

    .. code-block:: yaml

        min_pool_size: 1
        endpoints:
          # see HttpEndpoint
          - ...
        probe:
          # see HttpProbe
          ...

    """

    min_pool_size: int = 1
    endpoints: list[RemoteEndpointSettings] = None
    probe: Optional[RemoteProbeSettings] = None

    #: Events triggering the circuit breaker.
    break_on: set = {BREAK_ON_NETWORK_ERROR, BREAK_ON_UNHANDLED_EXCEPTION}

    #: Delay after which endpoints that are marked as down will be checked again.
    check_after: Decimal = 10.0

    def __getitem__(self, item):
        item = normalize_url(item)
        for endpoint in self.endpoints:
            if str(endpoint.url) == item:
                return endpoint
        raise KeyError(f'Endpoint "{item}" not found.')

    @field_validator("break_on")
    @classmethod
    def validate_break_on(cls, value: set) -> set:
        if not value.issubset(ALL_BREAK_ON_VALUES):
            raise ValueError(f"Invalid break_on values: {value}")
        return value

    @field_serializer("break_on", when_used="json")
    @classmethod
    def serialize_in_order(cls, value: Iterable[str]):
        return sorted(value)

    """
    XXX: This is a work in progress


    current_pool: Deque[RemoteEndpointSettings] | None = None
    current_pool_name = None
    def __init__(
        self,
        endpoints: Iterable[HttpEndpoint | dict | str] | str = (),
        *,
        min_pool_size=None,
        probe: Optional[HttpProbe | dict] = None,
        break_on: Optional[set] = None,
    ):
        # endpoints
        self.endpoints = {}
        self.current_pool_name = DEFAULT_POOL
        self.current_pool = deque()

        if isinstance(endpoints, str):
            endpoints = [endpoints]
        for endpoint in endpoints:
            if isinstance(endpoint, str):
                endpoint = {"url": endpoint}
            if not isinstance(endpoint, HttpEndpoint):
                endpoint = HttpEndpoint(**endpoint)
            self.endpoints.setdefault(endpoint.url, endpoint)
            if not self.endpoints[endpoint.url].pools:
                self.endpoints[endpoint.url].pools.add(DEFAULT_POOL)

        # min pool size
        self.min_pool_size = max(0, min_pool_size or self.min_pool_size)

        # probe
        if isinstance(probe, dict):
            probe = HttpProbe(**probe)
        self.probe = probe

        # break on
        break_on = set(break_on) if break_on is not None else self.break_on
        if not break_on.issubset(ALL_BREAK_ON_VALUES):
            raise ValueError(f"Invalid break_on values: {break_on}")
        self.break_on = break_on

        # compute first pool
        self.refresh()






    def _asdict(self, /, *, secure=True, with_status=False):
        return {
            "endpoints": [
                endpoint._asdict(secure=secure, with_status=with_status) for endpoint in self.endpoints.values()
            ],
            **({"min_pool_size": self.min_pool_size} if self.min_pool_size != type(self).min_pool_size else {}),
            **({"probe": self.probe._asdict(secure=secure)} if self.probe else {}),
            **({"current_pool": list(x.url for x in self.current_pool)} if with_status else {}),
            **({"break_on": list(sorted(self.break_on))} if self.break_on != type(self).break_on else {}),
        }

"""
