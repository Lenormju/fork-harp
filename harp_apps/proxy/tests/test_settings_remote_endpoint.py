import pytest
from pydantic import ValidationError

from harp.config.asdict import asdict
from harp.utils.collections import all_combinations
from harp.utils.testing.config import BaseConfigurableTest
from harp_apps.proxy.constants import AVAILABLE_POOLS
from harp_apps.proxy.settings.remote import RemoteEndpointSettings
from harp_apps.proxy.settings.remote.endpoint import RemoteEndpoint


class TestRemoteEndpointSettings(BaseConfigurableTest):
    type = RemoteEndpointSettings
    initial = {"url": "http://example.com"}
    expected = {"url": "http://example.com/"}
    expected_verbose = {
        "url": "http://example.com/",
        "pools": ["default"],
        "failure_threshold": 1,
        "success_threshold": 1,
    }

    def test_defaults2(self):
        settings = RemoteEndpointSettings(url="http://example.com")

        assert settings.model_dump(mode="json") == {
            "url": "http://example.com/",
            "pools": ["default"],
            "failure_threshold": 1,
            "success_threshold": 1,
        }

    @pytest.mark.parametrize("pools", all_combinations(AVAILABLE_POOLS))
    def test_valid_pools(self, pools):
        obj = self.create(pools=pools)
        assert asdict(obj) == {**self.expected, "pools": list(sorted(pools))}

    def test_invalid_pool(self):
        invalid_data = {"url": "http://example.com", "pools": ["invalid"]}
        with pytest.raises(ValueError):
            RemoteEndpointSettings(**invalid_data)

        with pytest.raises(ValidationError) as exc_info:
            RemoteEndpointSettings.model_validate(invalid_data)

        errors = exc_info.value.errors()
        assert errors[0]["loc"] == ("pools",)
        assert errors[0]["msg"] == "Value error, Invalid pool names: invalid."

    def test_valid_thresholds(self):
        valid_data = {
            "url": "http://example.com",
            "failure_threshold": 3,
            "success_threshold": 5,
        }
        settings = RemoteEndpointSettings(**valid_data)
        assert settings.failure_threshold == 3
        assert settings.success_threshold == 5

    def test_asdict(self):
        endpoint = self.create(url="http://example.com")
        expected = {
            "url": "http://example.com/",
        }
        assert asdict(endpoint) == expected

        # idempotence
        endpoint = self.create(**asdict(endpoint))
        assert asdict(endpoint) == expected

    def test_asdict_with_nondefault_thresholds(self):
        endpoint = self.create(url="http://example.com", success_threshold=2, failure_threshold=4)
        expected = {
            "url": "http://example.com/",
            "success_threshold": 2,
            "failure_threshold": 4,
        }
        assert asdict(endpoint) == expected

        # idempotence
        endpoint = RemoteEndpointSettings(**asdict(endpoint))
        assert asdict(endpoint) == expected


class TestRemoteEndpointStateful(BaseConfigurableTest):
    type = RemoteEndpoint
    initial = {"settings": TestRemoteEndpointSettings.initial}
    expected = {
        "settings": {
            "failure_threshold": 1,
            "pools": ["default"],
            "success_threshold": 1,
            "url": "http://example.com/",
        }
    }
    expected_verbose = {
        "settings": TestRemoteEndpointSettings.expected_verbose,
        "failure_reasons": None,
        "failure_score": 0,
        "status": 0,
        "success_score": 0,
    }
