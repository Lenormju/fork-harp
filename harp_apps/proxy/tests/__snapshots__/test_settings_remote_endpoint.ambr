# serializer version: 1
# name: TestRemoteEndpointSettings.test_jsonschema_for_serialization
  dict({
    'additionalProperties': False,
    'properties': dict({
      'failure_threshold': dict({
        'default': 1,
        'exclusiveMinimum': 0,
        'title': 'Failure Threshold',
        'type': 'integer',
      }),
      'pools': dict({
        'default': list([
          'default',
        ]),
        'items': dict({
          'type': 'string',
        }),
        'title': 'Pools',
        'type': 'array',
        'uniqueItems': True,
      }),
      'success_threshold': dict({
        'default': 1,
        'exclusiveMinimum': 0,
        'title': 'Success Threshold',
        'type': 'integer',
      }),
      'url': dict({
        'format': 'uri',
        'maxLength': 2083,
        'minLength': 1,
        'title': 'Url',
        'type': 'string',
      }),
    }),
    'required': list([
      'url',
    ]),
    'title': 'RemoteEndpointSettings',
    'type': 'object',
  })
# ---
# name: TestRemoteEndpointSettings.test_jsonschema_for_validation
  dict({
    'additionalProperties': False,
    'properties': dict({
      'failure_threshold': dict({
        'default': 1,
        'exclusiveMinimum': 0,
        'title': 'Failure Threshold',
        'type': 'integer',
      }),
      'pools': dict({
        'default': list([
          'default',
        ]),
        'items': dict({
          'type': 'string',
        }),
        'title': 'Pools',
        'type': 'array',
        'uniqueItems': True,
      }),
      'success_threshold': dict({
        'default': 1,
        'exclusiveMinimum': 0,
        'title': 'Success Threshold',
        'type': 'integer',
      }),
      'url': dict({
        'format': 'uri',
        'maxLength': 2083,
        'minLength': 1,
        'title': 'Url',
        'type': 'string',
      }),
    }),
    'required': list([
      'url',
    ]),
    'title': 'RemoteEndpointSettings',
    'type': 'object',
  })
# ---
# name: TestRemoteEndpointStateful.test_jsonschema_for_serialization
  dict({
    '$defs': dict({
      'RemoteEndpointSettings': dict({
        'additionalProperties': False,
        'description': '''
          A ``HttpEndpoint`` is an instrumented target URL that a proxy will use to route requests. It is used as the
          configuration parser for ``proxy.endpoints[].remote.endpoints[]`` settings.
          
          .. code-block:: yaml
          
              url: "http://my-endpoint:8080"
              pools: [fallback]  # omit for default pool
              failure_threshold: 3
              success_threshold: 1
        ''',
        'properties': dict({
          'failure_threshold': dict({
            'default': 1,
            'exclusiveMinimum': 0,
            'title': 'Failure Threshold',
            'type': 'integer',
          }),
          'pools': dict({
            'default': list([
              'default',
            ]),
            'items': dict({
              'type': 'string',
            }),
            'title': 'Pools',
            'type': 'array',
            'uniqueItems': True,
          }),
          'success_threshold': dict({
            'default': 1,
            'exclusiveMinimum': 0,
            'title': 'Success Threshold',
            'type': 'integer',
          }),
          'url': dict({
            'format': 'uri',
            'maxLength': 2083,
            'minLength': 1,
            'title': 'Url',
            'type': 'string',
          }),
        }),
        'required': list([
          'url',
        ]),
        'title': 'RemoteEndpointSettings',
        'type': 'object',
      }),
    }),
    'properties': dict({
      'failure_reasons': dict({
        'anyOf': list([
          dict({
            'items': dict({
            }),
            'type': 'array',
            'uniqueItems': True,
          }),
          dict({
            'type': 'null',
          }),
        ]),
        'default': None,
        'title': 'Failure Reasons',
      }),
      'failure_score': dict({
        'default': 0,
        'title': 'Failure Score',
        'type': 'integer',
      }),
      'settings': dict({
        '$ref': '#/$defs/RemoteEndpointSettings',
      }),
      'status': dict({
        'default': 0,
        'title': 'Status',
        'type': 'integer',
      }),
      'success_score': dict({
        'default': 0,
        'title': 'Success Score',
        'type': 'integer',
      }),
    }),
    'required': list([
      'settings',
    ]),
    'title': 'RemoteEndpoint',
    'type': 'object',
  })
# ---
# name: TestRemoteEndpointStateful.test_jsonschema_for_validation
  dict({
    '$defs': dict({
      'RemoteEndpointSettings': dict({
        'additionalProperties': False,
        'description': '''
          A ``HttpEndpoint`` is an instrumented target URL that a proxy will use to route requests. It is used as the
          configuration parser for ``proxy.endpoints[].remote.endpoints[]`` settings.
          
          .. code-block:: yaml
          
              url: "http://my-endpoint:8080"
              pools: [fallback]  # omit for default pool
              failure_threshold: 3
              success_threshold: 1
        ''',
        'properties': dict({
          'failure_threshold': dict({
            'default': 1,
            'exclusiveMinimum': 0,
            'title': 'Failure Threshold',
            'type': 'integer',
          }),
          'pools': dict({
            'default': list([
              'default',
            ]),
            'items': dict({
              'type': 'string',
            }),
            'title': 'Pools',
            'type': 'array',
            'uniqueItems': True,
          }),
          'success_threshold': dict({
            'default': 1,
            'exclusiveMinimum': 0,
            'title': 'Success Threshold',
            'type': 'integer',
          }),
          'url': dict({
            'format': 'uri',
            'maxLength': 2083,
            'minLength': 1,
            'title': 'Url',
            'type': 'string',
          }),
        }),
        'required': list([
          'url',
        ]),
        'title': 'RemoteEndpointSettings',
        'type': 'object',
      }),
    }),
    'properties': dict({
      'failure_reasons': dict({
        'anyOf': list([
          dict({
            'items': dict({
            }),
            'type': 'array',
            'uniqueItems': True,
          }),
          dict({
            'type': 'null',
          }),
        ]),
        'default': None,
        'title': 'Failure Reasons',
      }),
      'failure_score': dict({
        'default': 0,
        'title': 'Failure Score',
        'type': 'integer',
      }),
      'settings': dict({
        '$ref': '#/$defs/RemoteEndpointSettings',
      }),
      'status': dict({
        'default': 0,
        'title': 'Status',
        'type': 'integer',
      }),
      'success_score': dict({
        'default': 0,
        'title': 'Success Score',
        'type': 'integer',
      }),
    }),
    'required': list([
      'settings',
    ]),
    'title': 'RemoteEndpoint',
    'type': 'object',
  })
# ---
